core/domain/question_services.py

```diff
@@ -323,12 +323,12 @@ def update_skill_ids_of_questions(
 
 
 def get_question_summaries_and_skill_descriptions(
-        question_count, skill_ids, start_cursor):
+        question_skill_count, skill_ids, start_cursor):
     """Returns the list of question summaries linked to all the skills given by
     skill_ids.
 
     Args:
-        question_count: int. The number of question summaries to return.
+        question_skill_count: int. The number of question skill links to fetch.
```

Comment:
Not sure I understand this. Shouldn't the point of this function still be to return a bunch of questions for a given set of skills? Having multiple skills assigned to one question just means that your selection algorithm changes. But I would not expect the function signature to change.

Reply:
It does change because in the backend, we only get single question_skill_link for every count. It was fine before because there was only one skill for each question, but since now there are multiple ones, we have to increase the question_skill_count to keep the number of questions per page constant (at 10).

Reply:
Ah yes, I do see the issue with the querying. I can't think of anything better in terms of the algorithmic approach.

But I think, from the perspective of the caller of this function, this is a bit indirect. Presumably, the caller really wants to specify the number of questions -- having to specify the number of question/skill links and then truncate later is a bit of a fudge. Also, the caller is probably interested in the questions themselves and not the question/skill links.

So, how about having this return *at least* question_count questions instead? That gives the caller what it's really asking for (and perhaps some additional bonus questions), while keeping the question/skill links internal. The tricky part here is to handle the corner case where a question_id is repeated across a page boundary -- e.g. the entity just before the cursor and the entity just after may correspond to the same question ID. The documentation for this function needs to be very clear about what guarantees there are on the returned values if this function is called multiple times in sequence, and what the caller potentially needs to be careful about.

Reply:
Sorry I don't understand what do you mean by "return at least question_count questions". How should it be handled?

Reply:
If I understand correctly, there can be at most 3 question/skill links per question. By retrieving 30 question/skill links you are really retrieving 10-30 questions (depending on how many question/skill links each question has).

So, you could return the question IDs for those questions, instead of the question skill models. Presumably the caller actually wants the question IDs and not the question skill links.

Reply:
I think I was returning question_ids in the new function I wrote. I did deduplication there and returned the question_ids, but I think it makes sense that I shouldn't do deduplication in the storage layer.

If you mean it should only return 10 questions, there would be some problems with the cursor. Let's say the cursor is set to the 30th link, but the 10 questions only used 20 links, then the 20-30 links would never be returned anymore.

Please correct me if I misunderstood!

Reply:
Oh yes, sorry, my bad. It's just the input that should be phrased in terms of the (minimum) number of questions, rather than as a number of question/skill links. And I agree that you don't want to do deduplication in the storage layer; the storage layer stuff should be as simple as possible and shouldn't have much logic beyond the datastore calls/queries.

I think it's fine for this function to return 10+ questions, hence the suggestion to rephrase it as "get at least X questions". The caller can handle the surplus. But, as mentioned before, make sure to explain clearly in the docstring the potential pitfalls that can occur with regards to the independence (or not) of the results of successive calls. Tests should be written for those cases, too.

Reply:
I see. Just to make sure, what I need to do is changing the name from question_skill_count to minimum_question_count right? And the count would be 10 and we do minimum_question_count * 3 in the function? 

But it would be a bit inconvenient for skill_editor page because it only has one link per question, so we don't really need to take the complicated way. By doing this it's forced to use the hard way for both editors.

Reply:
Basically yes, that was my idea (for your first paragraph).

For the second paragraph, I actually think we should be unifying the code for fetching questions for a given set of skills into a single service in the frontend, and both frontend pages call that service (which handles all the complications internally). That would be better than having this functionality be repeated multiple times in multiple different pages (presumably it'll be needed later in the learner view, practice questions, etc. as well).

Reply:
Not done with this one yet.